File: Masjidly - Official Masjid Prayer Times/Data/Persistence/PrayerTimesDiskCache.swift
import Foundation

/// Atomic JSON file cache for prayer data stored in Application Support.
///
/// Layout:
/// ```
/// Application Support/PrayerTimesCache/
///   mosques.json
///   data_revision.json
///   uk_dst.json
///   monthly_{slug}_{month}_{year}.json
///   ramadan_{slug}_{date}.json
///   versions_{slug}_{month}_{year}.json
/// ```
@MainActor
final class PrayerTimesDiskCache: Sendable {
    private let fileManager: FileManager
    private let cacheDir: URL

    init(fileManager: FileManager = .default) {
        self.fileManager = fileManager
        let support = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        cacheDir = support.appendingPathComponent("PrayerTimesCache", isDirectory: true)
    }

    // MARK: - Helpers

    private func ensureCacheDir() throws {
        if !fileManager.fileExists(atPath: cacheDir.path) {
            try fileManager.createDirectory(at: cacheDir, withIntermediateDirectories: true)
        }
    }

    /// Sanitize a string so it is safe as a filename component (no path separators, no dots).
    static func safe(_ component: String) -> String {
        component.replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
            .replacingOccurrences(of: ".", with: "_")
    }

    private func url(for filename: String) -> URL {
        cacheDir.appendingPathComponent(filename, isDirectory: false)
    }

    /// Atomically write `data` to `url` via a temporary file + `replaceItemAt`.
    private func atomicWrite(data: Data, to url: URL) throws {
        let tmp = url.deletingLastPathComponent()
            .appendingPathComponent(".tmp_\(UUID().uuidString)", isDirectory: false)
        try data.write(to: tmp, options: .atomic)
        _ = try fileManager.replaceItemAt(url, withItemAt: tmp)
    }

    private func loadJSON<T: Decodable>(from url: URL) -> T? {
        guard fileManager.fileExists(atPath: url.path),
              let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }

    private func saveJSON<T: Encodable>(_ value: T, to url: URL) throws {
        try ensureCacheDir()
        let data = try JSONEncoder().encode(value)
        try atomicWrite(data: data, to: url)
    }

    // MARK: - Mosques

    private static let mosquesFile = "mosques.json"

    func loadMosques() -> [Mosque]? {
        loadJSON(from: url(for: Self.mosquesFile))
    }

    func saveMosques(_ mosques: [Mosque]) throws {
        try saveJSON(mosques, to: url(for: Self.mosquesFile))
    }

    // MARK: - Data revision

    private static let dataRevisionFile = "data_revision.json"

    func loadDataRevision() -> DataRevision? {
        loadJSON(from: url(for: Self.dataRevisionFile))
    }

    func saveDataRevision(_ revision: DataRevision) throws {
        try saveJSON(revision, to: url(for: Self.dataRevisionFile))
    }

    // MARK: - UK DST

    private static let ukDstFile = "uk_dst.json"

    func loadUkDst() -> UkDstCalendar? {
        loadJSON(from: url(for: Self.ukDstFile))
    }

    func saveUkDst(_ dst: UkDstCalendar) throws {
        try saveJSON(dst, to: url(for: Self.ukDstFile))
    }

    // MARK: - Versions

    private static func versionsFilename(slug: String, month: String, year: Int) -> String {
        "versions_\(Self.safe(slug))_\(month)_\(year).json"
    }

    func loadVersions(slug: String, month: String, year: Int) -> CachedPrayerDataVersions? {
        loadJSON(from: url(for: Self.versionsFilename(slug: slug, month: month, year: year)))
    }

    func saveVersions(slug: String, month: String, year: Int, versions: PrayerDataVersions) throws {
        try saveJSON(CachedPrayerDataVersions(versions: versions, checkedAt: Date()), to: url(for: Self.versionsFilename(slug: slug, month: month, year: year)))
    }

    // MARK: - Monthly

    private static func monthlyFilename(slug: String, month: String, year: Int) -> String {
        "monthly_\(Self.safe(slug))_\(month)_\(year).json"
    }

    func loadMonthly(slug: String, month: String, year: Int) -> MonthPrayerData? {
        loadJSON(from: url(for: Self.monthlyFilename(slug: slug, month: month, year: year)))
    }

    func saveMonthly(slug: String, month: String, year: Int, data: MonthPrayerData) throws {
        try saveJSON(data, to: url(for: Self.monthlyFilename(slug: slug, month: month, year: year)))
    }

    func removeMonthly(slug: String, month: String, year: Int) {
        let fileURL = url(for: Self.monthlyFilename(slug: slug, month: month, year: year))
        guard fileManager.fileExists(atPath: fileURL.path) else { return }
        try? fileManager.removeItem(at: fileURL)
    }

    // MARK: - Ramadan

    private static func ramadanFilename(slug: String, date: String) -> String {
        "ramadan_\(Self.safe(slug))_\(Self.safe(date)).json"
    }

    func loadRamadan(slug: String, date: String) -> RamadanPrayerData? {
        loadJSON(from: url(for: Self.ramadanFilename(slug: slug, date: date)))
    }

    func saveRamadan(slug: String, date: String, data: RamadanPrayerData) throws {
        try saveJSON(data, to: url(for: Self.ramadanFilename(slug: slug, date: date)))
    }
}

File: apps/android/app/src/main/kotlin/com/mikhailspeaks/masjidly/data/cache/PrayerTimesDiskCache.kt
package com.mikhailspeaks.masjidly.data.cache

import android.content.Context
import com.mikhailspeaks.masjidly.domain.CachedPrayerDataVersions
import com.mikhailspeaks.masjidly.domain.DataRevision
import com.mikhailspeaks.masjidly.domain.MonthPrayerData
import com.mikhailspeaks.masjidly.domain.Mosque
import com.mikhailspeaks.masjidly.domain.PrayerDataVersions
import com.mikhailspeaks.masjidly.domain.RamadanPrayerData
import com.mikhailspeaks.masjidly.domain.UkDstCalendar
import java.io.File
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * JSON file cache — mirrors iOS `PrayerTimesDiskCache.swift`.
 */
class PrayerTimesDiskCache(context: Context) {
    private val cacheDir = File(context.applicationContext.filesDir, "PrayerTimesCache").also { it.mkdirs() }
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    private fun file(name: String) = File(cacheDir, name)

    private fun safe(component: String): String =
        component.replace("/", "_").replace("\\", "_").replace(".", "_")

    private inline fun <reified T> load(name: String): T? {
        val f = file(name)
        if (!f.exists()) return null
        return runCatching { json.decodeFromString<T>(f.readText()) }.getOrNull()
    }

    private inline fun <reified T> save(name: String, value: T) {
        val target = file(name)
        val tmp = File(cacheDir, ".tmp_${System.nanoTime()}")
        tmp.writeText(json.encodeToString(value))
        if (target.exists()) target.delete()
        tmp.renameTo(target)
    }

    fun loadMosques(): List<Mosque>? =
        load<List<MosqueCache>>(MOSQUES_FILE)?.map { it.toDomain() }

    fun saveMosques(mosques: List<Mosque>) {
        save(MOSQUES_FILE, mosques.map { it.toCache() })
    }

    fun loadDataRevision(): DataRevision? =
        load<DataRevision>(DATA_REVISION_FILE)

    fun saveDataRevision(revision: DataRevision) {
        save(DATA_REVISION_FILE, revision)
    }

    fun loadUkDst(): UkDstCalendar? =
        load<UkDstCalendarCache>(UK_DST_FILE)?.toDomain()

    fun saveUkDst(dst: UkDstCalendar) {
        save(UK_DST_FILE, dst.toCache())
    }

    fun loadVersions(slug: String, month: String, year: Int): CachedPrayerDataVersions? =
        load<CachedPrayerDataVersions>(versionsFilename(slug, month, year))

    fun saveVersions(slug: String, month: String, year: Int, versions: PrayerDataVersions) {
        save(versionsFilename(slug, month, year), CachedPrayerDataVersions(versions, System.currentTimeMillis()))
    }

    fun loadMonthly(slug: String, month: String, year: Int): MonthPrayerData? {
        val name = monthlyFilename(slug, month, year)
        return load<MonthPrayerDataCache>(name)?.toDomain()
    }

    fun saveMonthly(slug: String, month: String, year: Int, data: MonthPrayerData) {
        save(monthlyFilename(slug, month, year), data.toCache())
    }

    fun removeMonthly(slug: String, month: String, year: Int) {
        file(monthlyFilename(slug, month, year)).delete()
    }

    fun loadRamadan(slug: String, date: String): RamadanPrayerData? {
        val name = ramadanFilename(slug, date)
        return load<RamadanPrayerDataCache>(name)?.toDomain()
    }

    fun saveRamadan(slug: String, date: String, data: RamadanPrayerData) {
        save(ramadanFilename(slug, date), data.toCache())
    }

    private fun monthlyFilename(slug: String, month: String, year: Int) =
        "monthly_${safe(slug)}_${month}_$year.json"

    private fun versionsFilename(slug: String, month: String, year: Int) =
        "versions_${safe(slug)}_${month}_$year.json"

    private fun ramadanFilename(slug: String, date: String) =
        "ramadan_${safe(slug)}_${safe(date)}.json"

    companion object {
        private const val MOSQUES_FILE = "mosques.json"
        private const val DATA_REVISION_FILE = "data_revision.json"
        private const val UK_DST_FILE = "uk_dst.json"
    }
}
